### ISSUE TITLE
Rewrite scheduler to avoid polling

### ISSUE THREAD

**Issue by Kinnay** [Kinnay/NintendoClients#36]
Currently, the scheduler loops through all sockets and timers every 20 milliseconds to check if they're ready. This is probably not scalable. Even if the performance impact is not as bad as it seems, it just feels bad to waste resources like that. Maybe we can use `select`?

**Comment by Kinnay** [comment]
There are other issues with the current scheduler. For example, imagine the following scenario:
1. Task A calls a blocking function, such as `PRUDPClient.connect`.
2. To prevent the scheduler from being stalled (i.e. to allow other tasks to keep running) blocking functions keep calling `scheduler.update` until they are done.
3.  Task B becomes ready while task A is still blocking, so it is executed within the `scheduler.update` call.
4. Task B calls another blocking function. The blocking function again keeps calling `scheduler.update` until it's done. However, this means that the `scheduler.update` call from task A will not return until task B is complete. Therefore, task A will be stuck until task B is complete, even if task A would otherwise be finished much earlier than task B.

I never noticed any issues with the current scheduler, but it's probably a good idea to think about this.

**Comment by Kinnay** [comment]
Asyncio looks pretty nice, but it seems like nothing can be done without asyncio anymore if a single function uses it.

I discovered [fibers](https://pypi.org/project/fibers/). This might be the way to go. The thread scheduler could create a fiber for each task internally. Fibers also seem much more lightweight than threads.

![image](https://user-images.githubusercontent.com/10852324/83946789-3aa5a280-a813-11ea-9ea8-fee514d87d99.png)

One question remains. The nice thing about the current implementation is that the prudp and websocket connections are managed in the background, so you can do whatever you like in your scripts without breaking the connection. To keep this, the scheduler has to run in a background thread, so I have to come up with a good way to interact with the background thread from the main thread. Also: #37.

**Comment by iomintz** [comment]
Asyncio can be used by sync clients if you make a synchronous wrapper which manages its own event loop and blocks on the futures using it. For example, here is a synchronous wrapper for an asynchronous database client library: https://github.com/iomintz/syncpg/blob/master/syncpg.py

Other than that, it is true that async propagates all the way up to the entry point of the process. This is good though: it means that any code that does non-blocking IO must be itself non-blocking. It also ensures that you can see exactly what function calls you make that are blocking.

Personally I would *love* for this library to be converted to asyncio (or anyio which would allow people to use alternative frameworks in addition to asyncio such as trio and curio). Currently this library is the only thing keeping me back from using Quart or aiohttp.web in my web server. And async is just so much nicer for the things I eventually want to do with it, like managing simultaneous requests.

Big vouch for asyncio (or anyio).

**Comment by Kinnay** [comment]
I don't have much experience with asyncio, but I feel like it takes away too much freedom. It's perfect if you're writing a server, but clients are generally much more interactive. For example, I can do something like this directly in my python shell:

![image](https://user-images.githubusercontent.com/10852324/83965113-89097e80-a8b1-11ea-8608-ddf78059be1c.png)

I don't think that would be possible anymore if I switched to asyncio. Everything would have to run in an event loop. I could use a synchronous wrapper like you suggested, but I would have to write such a wrapper around every single class that uses asyncio, which increases the code size and makes it hard to maintain.

I came across another client library that uses asyncio today: https://github.com/nats-io/nats.py. If you look at their examples you'll see `await` and `async` all over the place. *Everything* has to be written with asyncio. Basically, if I used asyncio I would force any scripts that use NintendoClients to use asyncio as well. I don't think I want that.

Maybe you can use something like this? https://stackoverflow.com/questions/22190403/how-could-i-use-requests-in-asyncio. It sounds like you can even run a synchronous function (like one from NintendoClients) asynchronously with asyncio.

**Comment by iomintz** [comment]
Yeah you have to use await and async everywhere. No that is not a problem. I have used many asynchronous clients before and haven't really encountered these issues. If you want an asynchronous REPL you can just use `python -m asyncio`. If you don't have python 3.8 you can use `aioconsole`'s included `apython` script. That will let you use await in the REPL. 

As for running synchronous code in an thread executor, we really only let that slide if it's CPU bound work that releases the GIL. If it's IO bound work it is always better to just find a way to do it in the event loop.

As for CLI programs, while I agree that it would be a bit more of a pain for those, it's not too hard to define an `amain()` and run it using asyncio.run(). If user input is needed it can be obtained in a thread using run_in_executor as there's no asyncio-based input() function. For sys.stdin, aiofiles can be used, I think. sys.stdout doesn't block long enough to care about that usually.

**Comment by Kinnay** [comment]
@iomintz I'm probably going to use explicit concurrency. I don't like the way asyncio deals with sockets though. In my opinion it would make much more sense if I could write something like `data = await sock.recv(size)`, but asyncio makes this really difficult. It does have functions like `loop.sock_recv`, but they don't work on SSL sockets and there's no `loop.sock_recvfrom` for UDP sockets.

I mostly ignored other async libraries earlier, but today I looked into them more thoroughly and I feel like curio has exactly what I need. What do you think?

**Comment by iomintz** [comment]
Have you looked at this: https://gist.github.com/zapstar/a7035795483753f7b71a542559afa83f? It doesn't seem too bad to me. If you want to use alternative async frameworks, I'd suggest [anyio](https://anyio.readthedocs.io/en/latest/networking.html), which lets your code run under asyncio, trio, and curio. If you must pick one, pick trio, but keep in mind this will *greatly* limit its compatibility as the trio ecosystem is not mature.

**Comment by Kinnay** [comment]
Yeah I saw the streams api, but it doesn't support UDP, which makes it kind of useless to me. I guess anyio could work, but its raw sockets seem to be hidden behind ugly SocketStream and UDPSocket classes. I might be able to use raw sockets with `anyio._get_asynclib().Socket`, but that's obviously not the intended way.

The nice thing about standard sockets is that you can easily switch between TCP and UDP sockets. If you want to switch to UDP you just have to pass a different socket type to `socket.socket()`. I don't get why all these async libraries want to place their own abstractions on top of sockets. I understand that TCP and UDP are different protocols, but that doesn't mean that you can't give them the same programming interface.

**Comment by iomintz** [comment]
Completely out of my league here, but don't you have to handle retrying and such yourself with UDP? They are different protocols, and UDP is more "hands-on" no?

**Comment by Kinnay** [comment]
Yes, UDP does not guarantee that packets arrive in the correct order or that they arrive at all. Of course you have to keep that in mind if you use UDP, but that's no reason to give UDP and TCP incompatible programming interfaces.

Right now I can easily change which protocol is used by PRUDP internally (see [prudp.py](https://github.com/Kinnay/NintendoClients/blob/master/nintendo/nex/prudp.py#L695-L702)), because my UDPClient and TCPClient provide the same interface. I really want to keep that.

My UDPClient and TCPClient are exactly the same (see [socket.py](https://github.com/Kinnay/NintendoClients/blob/master/nintendo/common/socket.py#L93-L132)) because python's sockets support both UDP and TCP with the same interface. The only reason that I have separate UDPClient and TCPClient classes is that my UDPServer and TCPServer classes are also separated and I wanted to make them more similar.

The whole point of an abstraction is to make the code simple and reusable.  In my opinion asyncio and anyio provide bad abstractions. `DatagramProtocol` is just another layer between the socket and your application code, but it doesn't feel abstract at all. I'd even say that it is less abstract than the socket itself, because sockets provide the same interface for UDP *and* TCP applications, while asyncio's protocol classes only provide an interface for UDP *or* TCP.

At the same time, it doesn't even make the code simpler. For example, compare this:
```python
import socket
import curio.io

async def main():
    s = socket.socket() # Create a python socket
    s = curio.io.Socket(s) # Create a curio socket
    await s.connect(("example.com", 80))
    await s.sendall(b"GET / HTTP/1.1\r\nHost: example.com\r\n\r\n")
    print(await s.recv(8192))
    await s.close()
curio.run(main)
```

Against this:
```python
class MyProtocol(asyncio.Protocol):
    def __init__(self):
        self.done = asyncio.Event()
    
    def connection_made(self, transport):
        self.transport = transport
        self.transport.write(b"GET / HTTP/1.1\r\nHost: example.com\r\n\r\n")
    
    def data_received(self, data):
        print(data)
        self.transport.close()
        self.done.set()

async def main():
    loop = asyncio.get_running_loop()
    transport, protocol = await loop.create_connection(MyProtocol, "example.com", 80)
    await protocol.done.wait()
asyncio.run(main())
```

This is way too much OOP for my taste and it's a pain to use. I have to create a whole new class just to use a socket, and receiving data is completely different from sending data.

In my view, asyncio's protocols only provide limitations instead of abstractions. Curio's Socket interface is much simpler, and also more general because it supports both TCP and UDP sockets.

**Comment by iomintz** [comment]
However you do it I implore you to maintain compatibility with asyncio.

**Comment by Kinnay** [comment]
Just dumping my thoughts here. I'm convinced that `async` and `await` are a good thing. It's still difficult to decide which library I want to use though. I've been playing around with all of them, but they all have something that I don't like.

* **asyncio:** ~I've read too many negative blog posts about asyncio.~ I don't really enjoy programming with asyncio for reasons that I explained above. Its documentation is also not always clear. For example, I'm still not sure what `connection_made` and `connection_lost` are supposed to mean in the UDP protocol. I'm slowing getting used to it though. It would just take some work to implement my own interface on top of asyncio.

* **curio:** Curio is easy and intuitive. But: it does not protect against multiple tasks accessing the same socket concurrently. For example, if one task blocks at a `recv` call and another tasks closes the socket the first task is stuck forever. I'd have to do a lot of bookkeeping to prevent this and there a lots of edge cases. I guess I could move the responsibility to the user, i.e. say that you should be careful not to close a socket while another task is using it, but it's easy to get this wrong.

* **trio:** Trio's low level socket api raises an exception if multiple tasks try to access a socket concurrently, which is really nice. However, its low level socket api does not support ssl sockets. If I want to use ssl I'm forced to use its `SSLStream`, which has a completely different interface for some reason (`close` is called `aclose`, `recv` is called `receive_from`, `sendall` is called `send_all`). Also, the only way to spawn a task in trio is with a nursery. I understand the reasoning behind this, but I'm not sure how I want to design my interfaces around this.

**Comment by iomintz** [comment]
With trio you can have a global nursery as a sort of "escape-hatch" or "opt-out" of the requirement to only spawn tasks inside of a nursery. You'll still have to handle recreating the nursery when a task fails though. Can you use anyio? I believe it exposes a similar interface to trio, and would enable you to support asyncio (which I would really prefer as I'd probably use aiohttp.web or Quart for my web server rewrite, assuming I don't use Elixir that is). Any differences in interface between SSL sockets and what have you can probably be abstracted away.

**Comment by Kinnay** [comment]
Ok I will probably use anyio. I kind of forgot about anyio before, but I looked into it again today and I'm starting to like it. I'm also becoming convinced that nurseries / task groups are a good thing. The file descriptor leak that you discovered happens because the sockets aren't closed properly. I also wouldn't be surprised if the background tasks aren't canceled properly and keep running forever. Both problems should go away if I use task groups and make sure that sockets are wrapped in an `async with` statement.

I just need to be careful, because the next anyio release will have some backwards compatibility breaking changes (they are removing queues for example).

**Comment by Kinnay** [comment]
I have a working implementation that uses anyio now. I pushed it to the 'anyio' branch, but I will probably move it to master soon.

@iomintz You seem to have more experience with async programming than me. Do you want to take a look? To see if anything jumps out you?

**Comment by iomintz** [comment]
I will take a look, however I don't have much experience with network programming. I have more experience using async libraries than writing them 😓

**Comment by Kinnay** [comment]
Thanks, I appreciate that. Don't worry about that. My only network programming experience comes from writing this library :wink:

**Comment by iomintz** [comment]
I'm not finished looking it over yet, but from a cursory glance I can see two issues:
1. use of `nintendo.common.http.request` prevents connection reuse. The previous code would instantiate a client  as an instance variable and reuse it.
2. I'm not sure exactly how the cryptography libraries are used, but one has to be careful using them lest they block the event loop for too long. If an operation is performed that blocks for too long (say, 50ms) it should be run in a worker thread.

**Comment by iomintz** [comment]
Why is a custom HTTP library used over something like [asks](https://pypi.org/project/asks)?

**Comment by Kinnay** [comment]
Thanks :)

> use of `nintendo.common.http.request` prevents connection reuse. The previous code would instantiate a client as an instance variable and reuse it.

I agree, connection reuse would be useful. When I tested nnas.py I noticed that setting up the TLS connection takes longer than the http request itself. The previous code didn't reuse the client either though. I did try to reuse connections at some point, but I removed that later because it had some issues (40fc0ca2).

> I'm not sure exactly how the cryptography libraries are used, but one has to be careful using them lest they block the event loop for too long. If an operation is performed that blocks for too long (say, 50ms) it should be run in a worker thread.

I don't think cryptographic operations take that long. I just measured it and a single RSA encryption takes less than a millisecond.



> Why is a custom HTTP library used over something like [asks](https://pypi.org/project/asks)?

That's a good question. It feels a bit like reinventing the wheel. I want to make sure that my requests look exactly like the requests that are sent by the real Switch, to reduce the risk of being banned. I want to precisely control the order of the headers for example. I wanted to use `requests`, but I remember that there was no way to remove a certain header from the request, which made it impossible to mimic the real Switch accurately. That's why I wrote my own HTTP client. I didn't look into asks, but since I already had my own HTTP client anyway it wasn't a lot of work to port it to anyio.
